proptest/src/test_runner/state_machine.rs

```diff
@@ -0,0 +1,193 @@
+//-
+// Copyright 2023 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Test declaration helpers and runners for abstract state machine testing.
+
+use super::Config;
+use crate::std_facade::Vec;
+use crate::strategy::state_machine::AbstractStateMachine;
+
+/// State machine test that relies on an abstract state machine model
+pub trait StateMachineTest {
```

Comment:
do we and if so, how do we want to think about testing async code? have you done any state machine testing with async code and if so, what was the experience like?